pootle/static/js/browser/components/TopContributors.js

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) Pootle contributors.
+ *
+ * This file is a part of the Pootle project. It is distributed under the GPL3
+ * or later license. See the LICENSE file for a copy of the license and the
+ * AUTHORS file for copyright and authorship information.
+ */
+
+import React from 'react';
+
+import Avatar from 'components/Avatar';
+import { t } from 'utils/i18n';
+
+
+const TopContributors = React.createClass({
+
+  propTypes: {
+    items: React.PropTypes.array.isRequired,
+    allItemsLoaded: React.PropTypes.bool.isRequired,
+    onLoadMore: React.PropTypes.func,
+  },
+
+  getInitialState() {
+    return {
+      items: this.props.items,
+      allItemsLoaded: this.props.allItemsLoaded,
+    };
+  },
+
+  loadMore() {
+    const _this = this;
+    this.props.onLoadMore((items, allItemsLoaded) => {
```

Comment:
This doesn't make much sense to me in terms of:

* Naming. `onLoadMore()` denotes something that will be done when N more items are loaded. Surprisingly, it's fetching the data.
* What's being done where. A callback is called to fetch the data AND the callback calls again the supplied callback to `setState`. Besides being twisted to follow, it feels wrong how data fetching happens, because the layer upwards is not a component.

A good read to clear up your mind on what state is and where should it go is the [Interactivity and Dynamic UIs](https://facebook.github.io/react/docs/interactivity-and-dynamic-uis.html) guide from the official docs.

Quoting from there:
> **Try to keep as many of your components as possible stateless**. By doing this you'll isolate the state to its most logical place and minimize redundancy, making it easier to reason about your application.
>
> A common pattern is to create several stateless components that just render data, and have a stateful component above them in the hierarchy that passes its state to its children via props. The stateful component encapsulates all of the interaction logic, while the stateless components take care of rendering data in a declarative way.

Following that suggestion, I'd have a `TopContributors` component which renders the given user data, and a `Stats` component above which will be responsible of holding the state (contribution data) and passing it down. Whenever the `Stats` component fetches new data and `setState`s, the `TopContributors` table will automatically update.
Note how having a `Stats` component above allows us to extend it to include more child components, so we can easily enhance it to render the rest of the items in the expanded stats view (which we already do to some extent). Implementing interactions becomes easier.
On the other hand, the components on the leaf nodes are very predictable since they are basically pure (i.e. given the same props, they provide the same output), and these are easy to test too and require no mocking at all.